### PR TITLE
[UpdateWorkflow] Ensure clustermgtd runs after cluster update and fix race condition making compute node deploy wrong cluster config version on update failure.

### PR DESCRIPTION
**Pull request by gmarciani**
### Description of changes
Ensure clustermgtd runs after cluster update (unconditionally on success, safely on failure) and fix race condition where compute nodes could deploy the wrong cluster config version after an update failure.

#### Update success
clustermgtd is restarted unconditionally at the end of the update recipe, regardless of whether the update includes queue changes.

#### Update or rollback failure
- Restart clustermgtd only if scontrol reconfigure succeeded, ensuring cluster management resumes safely
- Clean up DNA files shared with compute nodes to prevent them from deploying a config version that is about to be rolled back

#### Logs emitted by the handler

```
[2025-12-11T22:17:49+00:00] ERROR: Running exception handlers
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler: Started
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler: Update failed on HeadNode due to: execute[Check cluster readiness] (aws-parallelcluster-slurm::update_head_node line 169) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'

... omitted details about the error ...

[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler: Resources that have been successfully executed before the failure:
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler:   - ruby_block[Configure environment variable for recipes context: PATH]
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler:   - fetch_dna_files[Fetch ComputeFleet's Dna files]
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler:   - fetch_config[Fetch and load cluster configs]
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler:   - execute[stop clustermgtd]
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler:   - ruby_block[replace slurm queue nodes]
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler:   - execute[generate_pcluster_slurm_configs]
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler:   - execute[generate_pcluster_fleet_config]
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler:   - service[slurmctld]
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler:   - chef_sleep[5]
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler:   - execute[check slurmctld status]
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler:   - execute[reload config for running nodes]
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler:   - chef_sleep[15]
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler: Resource 'reload config for running nodes' has execution status: updated
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler: Running recovery commands
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler: Executing: cleanup DNA files
[2025-12-11T22:17:49+00:00] INFO: UpdateFailureHandler: Running command (attempt 1/11): /opt/parallelcluster/pyenv/versions/3.12.11/envs/cookbook_virtualenv/bin/python /opt/parallelcluster/scripts/share_compute_fleet_dna.py --region us-east-1 --cleanup
[2025-12-11T22:17:50+00:00] INFO: UpdateFailureHandler: Command stdout:
[2025-12-11T22:17:50+00:00] INFO: UpdateFailureHandler: Command stderr: INFO:__main__:Cleaning up /opt/parallelcluster/shared/dna/extra.json
INFO:__main__:Cleaning up /opt/parallelcluster/shared/dna/LaunchTemplateB1b67670b4d707d5-dna.json
INFO:__main__:All dna.json files have been shared!

[2025-12-11T22:17:50+00:00] INFO: UpdateFailureHandler: Successfully executed: cleanup DNA files
[2025-12-11T22:17:50+00:00] INFO: UpdateFailureHandler: scontrol reconfigure succeeded, starting clustermgtd
[2025-12-11T22:17:50+00:00] INFO: UpdateFailureHandler: Executing: start clustermgtd
[2025-12-11T22:17:50+00:00] INFO: UpdateFailureHandler: Running command (attempt 1/11): /opt/parallelcluster/pyenv/versions/3.12.11/envs/cookbook_virtualenv/bin/supervisorctl start clustermgtd
[2025-12-11T22:17:52+00:00] INFO: UpdateFailureHandler: Command stdout: clustermgtd: started

[2025-12-11T22:17:52+00:00] INFO: UpdateFailureHandler: Command stderr: /opt/parallelcluster/pyenv/versions/3.12.11/envs/cookbook_virtualenv/lib/python3.12/site-packages/supervisor/options.py:13: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  import pkg_resources

[2025-12-11T22:17:52+00:00] INFO: UpdateFailureHandler: Successfully executed: start clustermgtd
[2025-12-11T22:17:52+00:00] INFO: UpdateFailureHandler: Completed successfully
  - UpdateChefError::UpdateFailureHandler
Running handlers complete
```

### Tests
* Manual test where I injected failures in the update and rollback and verified that clustermgtd was running, no dna json file were left over.
* Manual test where I manually killed clustermgtd to simulate a corner case where it is not restarted, then verified that even an update with not changes to queues is able to restart it.
* Integ Test PENDING

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cookbooks/aws-parallelcluster-entrypoints/recipes/update.rb [hunk 1]
```diff
@@ -11,6 +11,12 @@
 # or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
 # OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
 # limitations under the License.
+
+chef_handler 'UpdateChefError::UpdateFailureHandler' do
+  type exception: true
+  action :enable
```
Comment (hanwen-cluster) [inline cookbooks/aws-parallelcluster-entrypoints/recipes/update.rb]: Minor, `action :enable` is the default. This line is optional

File: cookbooks/aws-parallelcluster-entrypoints/libraries/update_failure_handler.rb [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    DEFAULT_RETRIES = 10
+    DEFAULT_RETRY_DELAY = 90
+    DEFAULT_TIMEOUT = 30
+
+    def report
+      extend Chef::Mixin::ShellOut
+
+      Chef::Log.info("#{LOG_PREFIX} Started")
+
+      unless node_type == 'HeadNode'
```
Comment (hanwen-cluster) [inline cookbooks/aws-parallelcluster-entrypoints/libraries/update_failure_handler.rb:32]: Minor: if the logic is only about head node, the file name/class name could mention head node for better clarity

File: cookbooks/aws-parallelcluster-entrypoints/libraries/update_failure_handler.rb [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    def node_type
+      cluster_attributes['node_type']
+    end
+
+    def cookbook_virtualenv_path
+      "#{cluster_attributes['system_pyenv_root']}/versions/#{cluster_attributes['python-version']}/envs/cookbook_virtualenv"
+    end
+
+    def scontrol_reconfigure_succeeded?
+      slurm_reconfigure_status == :updated
```
Comment (hanwen-cluster) [inline cookbooks/aws-parallelcluster-entrypoints/libraries/update_failure_handler.rb]: minor: should we consider status "up_to_date"?

**Review by hanwen-cluster — DISMISSED** [review]

File: cookbooks/aws-parallelcluster-entrypoints/libraries/update_failure_handler.rb (hunk 2 above)
Comment (gmarciani) [reply cookbooks/aws-parallelcluster-entrypoints/libraries/update_failure_handler.rb:32]: **DESIGN PATTERNS DISCUSSION**

Good point. Your comment is much more than a minor about renaming a class; it actually opens an interesting discussion about design and how to enforce single responsibilities+open/close principle. :-) 

The solution that I have in mind is based on the strategy pattern, where entrypoints::update calls an UpdateFailureHandler, which is in charge of applying the right recovery strategy according to the node type. We do not want to bubble up this responsibility to upstream levels.

Regarding the name of the class, we should rename it only if :
1. the class *cannot be executed on node types != HeadNode*: this is not true, it can safely be executed on compute/login nodes, but for those nodes the recovery strategy at the moment is a no-op. In future we may need to introduce recovery strategies also for the other types of nodes.
2. the class has a single responsibility for the head node. I don't think the handler should be focused on the head node only. It's responsibility of the handler to decide the right strategy according to the node type (strategy pattern). Ideally, we should not even modify the handler body, but simply add a strategy for the specific node type to it and encapsulate that logic in the strategy. If we rename the class to something node type specific, then you would transfer the responsibility to decide the strategy to the entrypoint::update recipe, which is not correct (it would be equivalent to let that entrypoint decide which node-specific update recipe to apply, which is not the case).

In terms of single responsibilities, your comment makes me realize that the logic of command executions should be encapsulated into a dedicated class, for better separation of responsibilities. I'll do that.

what do you think?

**Review by hanwen-cluster — DISMISSED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -9,6 +9,9 @@ This file is used to list changes made in each version of the AWS ParallelCluste
 3.14.1
 ------
 
+**ENHANCEMENTS**
+- Ensure clustermgtd runs after cluster update. On success, restart unconditionally. On failure, restart if the queue reconfiguration succeeded.
```
Comment (hanwen-cluster) [inline CHANGELOG.md]: Is it `restart` or `start`?

File: cookbooks/aws-parallelcluster-entrypoints/libraries/update_failure_handler.rb (hunk 2 above)
Comment (hanwen-cluster) [reply cookbooks/aws-parallelcluster-entrypoints/libraries/update_failure_handler.rb:32]: I agree! Thank you!

File: CHANGELOG.md (hunk 4 above)
Comment (gmarciani) [reply CHANGELOG.md]: The operation we do is a `start`. I said "restart" because mostly we expect it to get stopped when the update begins.

Anyway, I will fix to `start` so that we are actually aligned to the real operation

**Review by hanwen-cluster — APPROVED** [review]
